libnavigation-core/src/main/java/com/mapbox/navigation/core/HistoryRecordingStateHandler.kt

```diff
@@ -82,15 +95,19 @@ internal class HistoryRecordingStateHandler(
     ) {
         val newState = NavigationSessionUtils.getNewState(isDriving, hasRoutes)
         if (newState::class != currentState::class) {
-            if (currentState !is NavigationSessionState.Idle) {
-                historyRecordingStateChangeObservers.forEach {
-                    finishRecordingBlock(it, currentState)
+            val oldState = currentState
+            currentState = newState
+            scope.launch {
```

Comment:
The scope we pass has `Main` dispatcher, which means that the coroutine won't be executed immediately, will it? Is it a problem for this feature?

Reply:
Hmm, I think it can be a problem, yes. If the observer receives `onShouldStartRecording` too late, we can miss `SET_ROUTES` event but that's exactly why this `HistoryRecordingStateChangedObserver` was introduced.
Great catch!
I even think that using immediate dispatcher is not enough here. Because `emit` can suspend. And we can then proceed with setting routes and get the new `DeveloperMetadata` after the `SET_ROUTES` event. I'm not sure if that's a problem though. @Guardiola31337 wdyt? How urgent do these notifications need to be?

Reply:
BTW, previously written instrumentation tests caught the problem you mentioned.